Implement mechanism to serve files via custom protocols

hmil opened the pull request:
In some scenarios, one may need the runtime files to be served over http:// instead of file:// (ie. to bypass cross origin request security).

In this first attempt, I created a new trait for file materializer which also provide an absolute access url for materialized files. This allows to encapsulate the "absolutePath" function together with the materializer. This trait has two direct implementations. One which is a refactoring of the current code (for the scenario where you don't care how selenium gets its runtime files, you just want it to work). The other one allows to customize the location where files get saved and the actual URL used by selenium to access them.

At usage it is pretty nice:

```
jsSettings(
  jsEnv := new SeleniumJSEnv(org.scalajs.jsenv.selenium.Firefox)
      // materializes files in the .tmp directory which is served over http by an external program
      .withMaterializer(new SpecificFileMaterializer(".tmp", "http://localhost:8080"))
)
```

The main problem I see with this approach is that my FileMaterializer trait is pretty redundant with respect to the VirtualFileMaterializer class. I suppose a cleaner approach would be to push all of this logic upstream such that VirtualFileMaterializer is the trait with two children and it exposes the getAbsolutePath method.

nicolasstucki commented:
You will need to add some tests that can be run on the CI.

nicolasstucki commented on seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/DefaultFileMaterializer.scala:
```diff
@@ -0,0 +1,18 @@
+
+/**
+  * Materializes virtual files in a temporary directory and links to them
+  * via file://
```
This comment should be formatted as follows.

``` scala
/** Materializes virtual files in a temporary directory and links to them
 *  via file://
 */
```

nicolasstucki commented on seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/FileMaterializer.scala:
```diff
@@ -0,0 +1,11 @@
+
+import org.scalajs.core.tools.io.VirtualTextFile
+
+
```
Remove the second new line.

nicolasstucki commented on seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SpecificFileMaterializer.scala:
```diff
@@ -0,0 +1,61 @@
+  *         .withMaterializer(new SpecificFileMaterializer(".tmp", "http://localhost:8080"))
+  * )
+  * </pre>
+  */
```
Same format issue with comment.

nicolasstucki commented on seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SpecificFileMaterializer.scala:
```diff
@@ -0,0 +1,61 @@
+  override def getAbsolutePath(fileName: String): String =
+    webRoot + "/" + fileName
+}
+
```
Put only one new line at the end.

nicolasstucki commented:
The code logic looks good to me, but we really need to test it on the CI.
That's all.

hmil commented:
Not sure how to test the feature exactly. I added a new target seleniumJSHttpEnv using the SpecificFileMaterializer and running one test which makes sure the script is served over HTTP.
I also added a python command to serve the files over http on travis.

I noticed there is no '.idea/' in the .gitignore, shall this be added ?

sjrd commented:
> I noticed there is no '.idea/' in the .gitignore, shall this be added ?

No. You should add that to your _global_ `.gitignore` file, so that it applies only on machine, but to all your projects.

sjrd commented on seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/DefaultFileMaterializer.scala:
```diff
@@ -0,0 +1,17 @@
+
+  private val materializer = new VirtualFileMaterializer(true)
+
+  override def materialize(vf: VirtualTextFile): Unit = materializer.materialize(vf)
```
I think, but needs verification, that it would be better if this method returned a `URL`, and `getAbsolutePath` was removed.

sjrd commented on seleniumJSHttpEnvTest/src/test/scala/org/scalajs/jsenv/selenium/LocationTest.scala:
```diff
@@ -0,0 +1,15 @@
+
+  @Test def LocationTest(): Unit = {
+    assertEquals(global.window.location.protocol.toString(), "http:")
+    assertEquals(global.window.location.host.toString(), "localhost:8080")
```
The expected value is traditionally on the left.

sjrd commented on seleniumJSHttpEnvTest/src/test/scala/org/scalajs/jsenv/selenium/LocationTest.scala:
```diff
@@ -0,0 +1,15 @@
+import org.junit.Test
+
+import scala.scalajs.js.Dynamic.global
+import scala.scalajs.js.isUndefined
```
Unused.

nicolasstucki commented on .travis.yml:
```diff
@@ -3,11 +3,13 @@ language: scala
+  - sbt ++$TRAVIS_SCALA_VERSION seleniumJSEnv/scalastyle seleniumJSEnv/test:scalastyle seleniumJSEnvTest/scalastyle seleniumJSHttpEnvTest/test:scalastyle seleniumJSEnvTest/test:scalastyle
   - sbt ++$TRAVIS_SCALA_VERSION seleniumJSEnvTest/run seleniumJSEnvTest/test 'set scalaJSStage in Global := FullOptStage' seleniumJSEnvTest/run seleniumJSEnvTest/test
   - sbt ++$TRAVIS_SCALA_VERSION 'set inScope(ThisScope in seleniumJSEnvTest)(jsEnv := new org.scalajs.jsenv.selenium.SeleniumJSEnv(org.scalajs.jsenv.selenium.Firefox).withKeepAlive())' seleniumJSEnvTest/run seleniumJSEnvTest/test 'set scalaJSStage in Global := FullOptStage' seleniumJSEnvTest/run seleniumJSEnvTest/test
+  - sbt ++$TRAVIS_SCALA_VERSION seleniumJSHttpEnvTest/test
```
You should add `fullOpt` to the tests:

```
- sbt ++$TRAVIS_SCALA_VERSION seleniumJSHttpEnvTest/test 'set scalaJSStage in Global := FullOptStage' seleniumJSHttpEnvTest/test
```

nicolasstucki commented on .travis.yml:
```diff
@@ -3,11 +3,13 @@ language: scala
   - "sh -e /etc/init.d/xvfb start"
-  - sleep 3 # give xvfb some time to start
+  - python -m SimpleHTTPServer 8080 &
+  - sleep 3
```
Given that the `scalastyle` took 165.06s in one of the travis test instanced, I would just ignore the `sleep` here.

nicolasstucki commented on build.sbt:
```diff
@@ -65,13 +75,16 @@ lazy val seleniumJSEnv: Project = project.
+  settings(testSettings).
+  settings(
+    jsEnv := new SeleniumJSEnv(Firefox).
+      withMaterializer(new CustomFileMaterializer(".tmp", "http://localhost:8080/.tmp"))
```
There is no real need for this folder to be hidden, you should just use `tmp`.

nicolasstucki commented on seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/CustomFileMaterializer.scala:
```diff
@@ -0,0 +1,57 @@
+  *         .withMaterializer(new SpecificFileMaterializer(".tmp", "http://localhost:8080"))
+  * )
+  * </pre>
+  */
```
Please follow our coding style for all comments (issue with allignment of `*`): 
https://github.com/scala-js/scala-js/blob/master/CODINGSTYLE.md#scaladoc

nicolasstucki commented on seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/CustomFileMaterializer.scala:
```diff
@@ -0,0 +1,57 @@
+  }
+
+  /**
+    * Creates the storage directory if it does not exist
```
And in this case the comment should also start on the first line.

nicolasstucki commented:
That is all.

nicolasstucki commented:
Could you also change the commit message to `Fix #22: Implement mechanism to serve files via custom protocols.` with no additional description.

hmil commented:
All done, thanks for the review.

nicolasstucki commented:
LGTM
